debrelease.1
.TH DEBRELEASE 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
debrelease \- a wrapper around dupload or dput
.SH SYNOPSIS
\fBdebrelease\fR [\fIdebrelease options\fR] [\fIdupload/dput options\fR]
.SH DESCRIPTION
\fBdebrelease\fR is a simple wrapper around \fBdupload\fR or
\fBdput\fR.  It is called from within the source code tree of a
package, and figures out the current version of a package.  It then
looks for the corresponding \fI.changes\fR file (which lists the files
needed to upload in order to release the package) in the parent
directory of the source code tree and calls \fBdupload\fR or
\fBdput\fR with the \fI.changes\fR file as parameter in order to
perform the actual uploading.
.PP
Options may be given to \fBdebrelease\fR; except for the ones listed
below, they are passed on unchanged to \fBdupload\fR or \fBdput\fR.
The \fBdevscripts\fR configuration files are also read by
\fBdebrelease\fR as described below.
.SH "Directory name checking"
In common with several other scripts in the \fBdevscripts\fR package,
\fBdebrelease\fR will climb the directory tree until it finds a
\fIdebian/changelog\fR file.  As a safeguard against stray files
causing potential problems, it will examine the name of the parent
directory once it finds the \fIdebian/changelog\fR file, and check
that the directory name corresponds to the package name.  Precisely
how it does this is controlled by two configuration file variables
DEVSCRIPTS_CHECK_DIRNAME_LEVEL and DEVSCRIPTS_CHECK_DIRNAME_REGEX, and
their corresponding command-line options \fB\-\-check-dirname-level\fR
and \fB\-\-check-dirname-regex\fR.
.PP
DEVSCRIPTS_CHECK_DIRNAME_LEVEL can take the following values:
.TP
.B 0
Never check the directory name.
.TP
.B 1
Only check the directory name if we have had to change directory in
our search for \fIdebian/changelog\fR.  This is the default behaviour.
.TP
.B 2
Always check the directory name.
.PP
The directory name is checked by testing whether the current directory
name (as determined by \fBpwd\fR(1)) matches the regex given by the
configuration file option DEVSCRIPTS_CHECK_DIRNAME_REGEX or by the
command line option \fB\-\-check-dirname-regex\fR \fIregex\fR.  Here
\fIregex\fR is a Perl regex (see \fBperlre\fR(3perl)), which will be
anchored at the beginning and the end.  If \fIregex\fR contains a '/',
then it must match the full directory path.  If not, then it must
match the full directory name.  If \fIregex\fR contains the string
\'PACKAGE', this will be replaced by the source package name, as
determined from the changelog.  The default value for the regex is:
\'PACKAGE(-.*)?', thus matching directory names such as PACKAGE and
PACKAGE-version.
.SH OPTIONS
.TP
\fB\-\-dupload\fR, \fB\-\-dput\fR
This specifies which uploader program to use; the default is
\fBdupload\fR.
.TP
\fB\-S\fR
If this option is used, or the default \fI.changes\fR file is
not found but a source-only \fI.changes\fR file is present, then this
source-only \fI.changes\fR file will be uploaded instead of an
arch-specific one.
.TP
\fB\-a\fIdebian-architecture\fR, \fB\-t\fIGNU-system-type\fR
See \fBdpkg-architecture\fR(1) for a description of these options.
They affect the search for the \fI.changes\fR file.  They are provided
to mimic the behaviour of \fBdpkg-buildpackage\fR when determining the
name of the \fI.changes\fR file.  If a plain \fB\-t\fR is given, it is
taken to be the \fBdupload\fR host-specifying option, and therefore
signifies the end of the \fBdebrelease\fR-specific options.
.TP
\fB\-\-multi\fR
Multiarch changes mode: This signifies that \fBdebrelease\fR should
use the most recent file with the name pattern
\fIpackage_version_*+*.changes\fR as the changes file, allowing for the
changes files produced by \fBdpkg-cross\fR.
.TP
\fB\-\-check-dirname-level\fR \fIN\fR
See the above section "Directory name checking" for an explanation of
this option.
.TP
\fB\-\-check-dirname-regex\fR \fIregex\fR
See the above section "Directory name checking" for an explanation of
this option.
.TP
\fB\-\-no-conf\fR, \fB\-\-noconf\fR
Do not read any configuration files.  This can only be used as the
first option given on the command-line.
.TP
.BR \-\-help ", " \-h
Display a help message and exit successfully.
.TP
.B \-\-version
Display version and copyright information and exit successfully.
.SH "CONFIGURATION VARIABLES"
The two configuration files \fI/etc/devscripts.conf\fR and
\fI~/.devscripts\fR are sourced in that order to set configuration
variables.  Command line options can be used to override configuration
file settings.  Environment variable settings are ignored for this
purpose.  The currently recognised variables are:
.TP
.B DEBRELEASE_UPLOADER
The currently recognised values are \fIdupload\fR and \fIdput\fR, and
it specifies which uploader program should be used.  It corresponds to
the \fB\-\-dupload\fR and \fB\-\-dput\fR command line options.
.TP
.BR DEVSCRIPTS_CHECK_DIRNAME_LEVEL ", " DEVSCRIPTS_CHECK_DIRNAME_REGEX
See the above section "Directory name checking" for an explanation of
these variables.  Note that these are package-wide configuration
variables, and will therefore affect all \fBdevscripts\fR scripts
which check their value, as described in their respective manpages and
in \fBdevscripts.conf\fR(5).
.SH "SEE ALSO"
.BR dupload (1),
.BR dput (1)
and
.BR devscripts.conf (5).
.SH AUTHOR
Julian Gilbey <jdg@debian.org>, based on the original \fBrelease\fR
script by Christoph Lameter <clameter@debian.org>.


debrelease.sh
#! /bin/bash

# debrelease: a devscripts wrapper around dupload/dput which calls
#             dupload/dput with the correct .changes file as parameter.
#             All command line options are passed onto dupload.
#
# Written and copyright 1999-2003 by Julian Gilbey <jdg@debian.org> 
# Based on the original 'release' script by
#  Christoph Lameter <clameter@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

set -e

PROGNAME=`basename $0`
MODIFIED_CONF_MSG='Default settings modified by devscripts configuration files:'

usage () {
    echo \
"Usage: $PROGNAME [debrelease options] [dupload/dput options]
  Run dupload on the newly created changes file.
  Debrelease options:
    --dupload         Use dupload to upload files (default)
    --dput            Use dput to upload files
    -a<arch>          Search for .changes file made for Debian build <arch>
    -t<target>        Search for .changes file made for GNU <target> arch
    -S                Search for source-only .changes file instead of arch one
    --multi           Search for multiarch .changes file made by dpkg-cross
    --check-dirname-level N
                      How much to check directory names before cleaning trees:
                      N=0   never
                      N=1   only if program changes directory (default)
                      N=2   always
    --check-dirname-regex REGEX
                      What constitutes a matching directory name; REGEX is
                      a Perl regular expression; the string \`PACKAGE' will
                      be replaced by the package name; see manpage for details
                      (default: 'PACKAGE(-.*)?')
    --no-conf, --noconf
                      Don't read devscripts config files;
                      must be the first option given
    --help            Show this message
    --version         Show version and copyright information

$MODIFIED_CONF_MSG"
}

version () {
    echo \
"This is $PROGNAME, from the Debian devscripts package, version ###VERSION###
This code is copyright 1999-2003 by Julian Gilbey, all rights reserved.
Based on original code by Christoph Lameter.
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 2 or later."
}

mustsetvar () {
    if [ "x$2" = x ]
    then
	echo >&2 "$PROGNAME: unable to determine $3"
	exit 1
    else
	# echo "$PROGNAME: $3 is $2"
	eval "$1=\"\$2\""
    fi
}

# Boilerplate: set config variables
DEFAULT_DEBRELEASE_UPLOADER=dupload
DEFAULT_DEVSCRIPTS_CHECK_DIRNAME_LEVEL=1
DEFAULT_DEVSCRIPTS_CHECK_DIRNAME_REGEX='PACKAGE(-.*)?'
VARS="DEBRELEASE_UPLOADER DEVSCRIPTS_CHECK_DIRNAME_LEVEL DEVSCRIPTS_CHECK_DIRNAME_REGEX"

if [ "$1" = "--no-conf" -o "$1" = "--noconf" ]; then
    shift
    MODIFIED_CONF_MSG="$MODIFIED_CONF_MSG
  (no configuration files read)"

    # set defaults
    for var in $VARS; do
	eval "$var=\$DEFAULT_$var"
    done
else
    # Run in a subshell for protection against accidental errors
    # in the config files
    eval $(
	set +e
	for var in $VARS; do
	    eval "$var=\$DEFAULT_$var"
	done

	for file in /etc/devscripts.conf ~/.devscripts
	  do
	  [ -r $file ] && . $file
	done

	set | egrep "^(DEBRELEASE|DEVSCRIPTS)_")

    # check sanity
    case "$DEBRELEASE_UPLOADER" in
	dupload|dput) ;;
	*) DEBRELEASE_UPLOADER=dupload ;;
    esac
    case "$DEVSCRIPTS_CHECK_DIRNAME_LEVEL" in
	0|1|2) ;;
	*) DEVSCRIPTS_CHECK_DIRNAME_LEVEL=1 ;;
    esac

    # set config message
    MODIFIED_CONF=''
    for var in $VARS; do
	eval "if [ \"\$$var\" != \"\$DEFAULT_$var\" ]; then
	    MODIFIED_CONF_MSG=\"\$MODIFIED_CONF_MSG
  $var=\$$var\";
	MODIFIED_CONF=yes;
	fi"
    done

    if [ -z "$MODIFIED_CONF" ]; then
	MODIFIED_CONF_MSG="$MODIFIED_CONF_MSG
  (none)"
    fi
fi


# synonyms
CHECK_DIRNAME_LEVEL="$DEVSCRIPTS_CHECK_DIRNAME_LEVEL"
CHECK_DIRNAME_REGEX="$DEVSCRIPTS_CHECK_DIRNAME_REGEX"


sourceonly=
multiarch=

while [ $# -gt 0 ]
do
    case "$1" in
    -a*) targetarch="`echo \"$1\" | sed -e 's/^-a//'`" ;;
    -t*) targetgnusystem="`echo \"$1\" | sed -e 's/^-t//'`"
	 # dupload has a -t option
	 if [ -z "$targetgnusystem" ]; then break; fi ;;
    -S) sourceonly=source ;;
    --multi) multiarch=yes ;;
    --dupload) DEBRELEASE_UPLOADER=dupload ;;
    --dput) DEBRELEASE_UPLOADER=dput ;;
    --check-dirname-level=*)
	level="`echo \"$1\" | sed -e 's/^--check-dirname-level=//'`"
        case "$level" in
	0|1|2) CHECK_DIRNAME_LEVEL=$level ;;
	*) echo "$PROGNAME: unrecognised --check-dirname-level value (allowed are 0,1,2)" >&2
	   exit 1 ;;
        esac
	;;
    --check-dirname-level)
	shift
        case "$1" in
	0|1|2) CHECK_DIRNAME_LEVEL=$1 ;;
	*) echo "$PROGNAME: unrecognised --check-dirname-level value (allowed are 0,1,2)" >&2
	   exit 1 ;;
        esac
	;;
    --check-dirname-regex=*)
	regex="`echo \"$1\" | sed -e 's/^--check-dirname-level=//'`"
	if [ -z "$regex" ]; then
	    echo "$PROGNAME: missing --check-dirname-regex parameter" >&2
	    echo "try $PROGNAME --help for usage information" >&2
	    exit 1
	else
	    CHECK_DIRNAME_REGEX="$regex"
	fi
	;;
    --check-dirname-regex)
	shift;
	if [ -z "$1" ]; then
	    echo "$PROGNAME: missing --check-dirname-regex parameter" >&2
	    echo "try $PROGNAME --help for usage information" >&2
	    exit 1
	else
	    CHECK_DIRNAME_REGEX="$1"
	fi
	;;
    --no-conf|--noconf)
	echo "$PROGNAME: $1 is only acceptable as the first command-line option!" >&2
	exit 1 ;;
    --dopts) shift; break ;;  # This is an option for cvs-debrelease,
                              # so we accept it here too, even though we don't
                              # advertise it
    --help) usage; exit 0 ;;
    --version) version; exit 0 ;;
    *) break ;;  # a dupload/dput option, so stop parsing here
    esac
    shift
done


# Look for .changes file via debian/changelog
CHDIR=
until [ -f debian/changelog ]; do
    CHDIR=yes
    cd ..
    if [ `pwd` = "/" ]; then
	echo "$PROGNAME: cannot find debian/changelog anywhere!" >&2
	echo "Are you in the source code tree?" >&2
	exit 1
    fi
done

mustsetvar package "`dpkg-parsechangelog | sed -n 's/^Source: //p'`" \
    "source package"
mustsetvar version "`dpkg-parsechangelog | sed -n 's/^Version: //p'`" \
    "source version"

if [ $CHECK_DIRNAME_LEVEL -eq 2 -o \
    \( $CHECK_DIRNAME_LEVEL -eq 1 -a "$CHDIR" = yes \) ]; then
    if ! perl -MFile::Basename -w \
	-e "\$pkg='$package'; \$re='$CHECK_DIRNAME_REGEX';" \
	-e '$re =~ s/PACKAGE/\\Q$pkg\\E/g; $pwd=`pwd`; chomp $pwd;' \
	-e 'if ($re =~ m%/%) { eval "exit (\$pwd =~ /^$re\$/ ? 0:1);"; }' \
	-e 'else { eval "exit (basename(\$pwd) =~ /^$re\$/ ? 0:1);"; }'
    then
	echo >&2 <<EOF
$progname: found debian/changelog for package $PACKAGE in the directory
  $pwd
but this directory name does not match the package name according to the
regex  $check_dirname_regex.

To run $progname on this package, see the --check-dirname-level and
--check-dirname-regex options; run $progname --help for more info.
EOF
	exit 1
    fi
fi


if [ "x$sourceonly" = "xsource" ]; then
    arch=source
else
    mustsetvar arch "`dpkg-architecture -a${targetarch} -t${targetgnusystem} -qDEB_HOST_ARCH`" "build architecture"
fi

sversion=`echo "$version" | perl -pe 's/^\d+://'`
pva="${package}_${sversion}_${arch}"
pvs="${package}_${sversion}_source"
changes="../$pva.changes"
schanges="../$pvs.changes"
mchanges=$(ls "../${package}_${sversion}_*+*.changes" "../${package}_${sversion}_multi.changes" 2>/dev/null | head -1)

if [ -n "$multiarch" ]; then
    if [ -z "$mchanges" -o ! -r "$mchanges" ]; then
	echo "$PROGNAME: could not find/read any multiarch .changes file with name" >&2
	echo "../${package}_${sversion}_*.changes" >&2
	exit 1
    fi
    changes=$mchanges
elif [ "$arch" = source ]; then
	if [ -r "$schanges" ]; then
	    changes=$schanges
	else
	    echo "$PROGNAME: could not find/read changes file $schanges!" >&2
	    exit 1
	fi
else
    if [ ! -r "$changes" ]; then
	if [ -r "$mchanges" ]; then
	    changes=$mchanges
	    echo "$PROGNAME: could only find a multiarch changes file:" >&2
	    echo "  $mchanges" >&2
	    echo -n "Should I upload this file? (y/n) " >&2
	    read ans
	    case ans in
		y*) ;;
		n*) exit 1 ;;
	    esac
	else
	    echo "$PROGNAME: could not read changes file $changes!" >&2
	    exit 1
	fi
    else
	echo "$PROGNAME: could not read changes file $changes!" >&2
	exit 1
    fi
fi

exec $DEBRELEASE_UPLOADER "$@" "$changes"

echo "$PROGNAME: failed to exec $DEBRELEASE_UPLOADER!" >&2
echo "Aborting...." >&2
exit 1
